manager/metrics/collector.go

```diff
@@ -0,0 +1,101 @@
+package metrics
+
+import (
+	"context"
+
+	"strings"
+
+	metrics "github.com/docker/go-metrics"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/manager/state/store"
+)
+
+var (
+	nodesMetric metrics.LabeledGauge
+)
+
+func init() {
+	ns := metrics.NewNamespace("swarmkit", "", nil)
+	nodesMetric = ns.NewLabeledGauge("nodes", "The number of nodes", metrics.Total, "state")
+	for _, state := range api.NodeStatus_State_name {
+		nodesMetric.WithValues(strings.ToLower(state)).Set(0)
+	}
+	metrics.Register(ns)
+}
+
+// Collector collects swarmkit metrics
+type Collector struct {
+	store *store.MemoryStore
+
+	// stopChan signals to the state machine to stop running.
+	stopChan chan struct{}
+	// doneChan is closed when the state machine terminates.
+	doneChan chan struct{}
+
+	nodesByState map[string]api.NodeStatus_State
+}
+
+// NewCollector creates a new metrics collector
+func NewCollector(store *store.MemoryStore) *Collector {
+	return &Collector{
+		store:        store,
+		stopChan:     make(chan struct{}),
+		doneChan:     make(chan struct{}),
+		nodesByState: make(map[string]api.NodeStatus_State),
+	}
+}
+
+func (c *Collector) updateNodeState(n *api.Node, remove bool) {
+	if prevState, ok := c.nodesByState[n.ID]; ok {
+		nodesMetric.WithValues(strings.ToLower(prevState.String())).Dec(1)
```

Comment:
`api.EventUpdateNode` may or may not change a node's state. You should check that before `Dec/Inc`.